crates/provider/src/layers/gas.rs

```diff
@@ -0,0 +1,207 @@
+use crate::{PendingTransactionBuilder, Provider, ProviderLayer, RootProvider};
+use alloy_network::{Network, TransactionBuilder};
+use alloy_primitives::U256;
+use alloy_transport::{Transport, TransportResult};
+use async_trait::async_trait;
+use std::marker::PhantomData;
+
+/// A layer that provides gas estimation for transactions.
+/// Populates the gas price and gas limit for transactions.
+#[derive(Debug, Clone, Copy)]
+pub struct GasEstimatorLayer;
+
+impl<P, N, T> ProviderLayer<P, N, T> for GasEstimatorLayer
+where
+    P: Provider<N, T>,
+    N: Network,
+    T: Transport + Clone,
+{
+    type Provider = GasEstimatorProvider<N, T, P>;
+    fn layer(&self, inner: P) -> Self::Provider {
+        GasEstimatorProvider { inner, _phantom: PhantomData }
+    }
+}
+
+/// A provider that provides gas estimation for transactions.
+#[derive(Debug, Clone)]
+pub struct GasEstimatorProvider<N, T, P>
+where
+    N: Network,
+    T: Transport + Clone,
+    P: Provider<N, T>,
+{
+    inner: P,
+    _phantom: PhantomData<(N, T)>,
+}
+
+impl<N, T, P> GasEstimatorProvider<N, T, P>
+where
+    N: Network,
+    T: Transport + Clone,
+    P: Provider<N, T>,
+{
+    async fn get_gas_price(&self) -> TransportResult<U256> {
+        let gas_price = self.inner.get_gas_price().await?;
+        Ok(gas_price)
+    }
+
+    async fn get_gas_estimate(&self, tx: &N::TransactionRequest) -> TransportResult<U256> {
+        let gas_estimate = self.inner.estimate_gas(tx, None).await?;
+        Ok(gas_estimate)
+    }
+
+    async fn get_eip1159_fees_estimate(&self) -> TransportResult<(U256, U256)> {
+        let (max_fee_per_gas, max_priority_fee_per_gas) =
+            self.inner.estimate_eip1559_fees(None).await?;
+
+        Ok((max_fee_per_gas, max_priority_fee_per_gas))
+    }
+}
+
+#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
+#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
+impl<N, T, P> Provider<N, T> for GasEstimatorProvider<N, T, P>
+where
+    N: Network,
+    T: Transport + Clone,
+    P: Provider<N, T>,
+{
+    fn root(&self) -> &RootProvider<N, T> {
+        self.inner.root()
+    }
+
+    async fn send_transaction(
+        &self,
+        mut tx: N::TransactionRequest,
+    ) -> TransportResult<PendingTransactionBuilder<'_, N, T>> {
+        let (gas_price, gas_limit, eip1559_fees) = match (
+            tx.gas_price().is_none(),
+            tx.gas_limit().is_none(),
+            (tx.max_fee_per_gas().is_none() && tx.max_priority_fee_per_gas().is_none()),
+        ) {
+            (true, true, true) => {
```

Comment:
pls add comments to make it clearer what tx types we are filling

Reply:
@onbjerg by tx types do you mean `EIP1559`, or `legacy` or do you mean specifying the fields being populated?

Reply:
for this, we can assume that there are 2 intended ways:
- 1559 style (assumed default)
- legacy style (IF and ONLY IF the user has already set the `gas_price` key).

the builder should have a type-level guarantee that `gas_price` and the 1559 gas fields are mutually exclusive, i.e. `(true,true,true)` should never be permitted at all.

as a result we should be able to simplify the 8 match arms to an `if let Some(price) = tx.gas_price` or similar

Reply:
in the future, we might want to allow a network-associated `GasStrategy` to handle this 🤔

Reply:
I don't quite understand this part though, if this is supposed to fill `gas_price` or `max_fee_per_gas`/`max_prio_fee_per_gas`, would they not always be unset? i.e. it would never be a legacy tx unless you set a gas price, in which case, we probably shouldn't override it?

Reply:
> @onbjerg by tx types do you mean EIP1559, or legacy or do you mean specifying the fields being populated?

i mean, please add comments on the arms to explain what we are matching for, it is quite hard to parse what `(true, true, false)` etc means

Reply:
> I don't quite understand this part though, if this is supposed to fill `gas_price` or `max_fee_per_gas`/`max_prio_fee_per_gas`, would they not always be unset? i.e. it would never be a legacy tx unless you set the `gas_price` field, in which case, we probably shouldn't override it?

Agree with @onbjerg's, point here. 

How about?
```rust
if tx.gas_price().is_none() {
   // Assume its a EIP1559 tx
   // Populate the following fields: gas_limit, max_fee_per_gas and max_priority_fee_per_gas 
} else {
   // Assume its a legacy tx
   // Populate only the gas_limit field if unset.
 }
```

Reply:
> > I don't quite understand this part though, if this is supposed to fill `gas_price` or `max_fee_per_gas`/`max_prio_fee_per_gas`, would they not always be unset? i.e. it would never be a legacy tx unless you set the `gas_price` field, in which case, we probably shouldn't override it?
> 
> Agree with @onbjerg's, point here.
> 
> How about?
> 
> ```rust
> if tx.gas_price().is_none() {
>    // Assume its a EIP1559 tx
>    // Populate the following fields: gas_limit, max_fee_per_gas and max_priority_fee_per_gas 
> } else {
>    // Assume its a legacy tx
>    // Populate only the gas_limit field if unset.
>  }
> ```

Yes, this is the correct strategy. Recommend decomposing into `fill_legacy` and `fill_1559` functions

Reply:
Pushed changes. 

In this update, if the `gas_price` is not set (i.e assumed to be 1559 tx) but the network doesn't support eip1559 txs then it falls back to being process as a legacy tx.